### DISCUSSION TITLE
Are there any items required in the context menu of the variable view?

### DISCUSSION THREAD

**Discussion by zero-plusplus** [zero-plusplus/vscode-autohotkey-debug#119] [Ideas]
Starting with version [1.49](https://code.visualstudio.com/updates/v1_49#_contributable-context-menu-for-variables-view) of vscode, can now customize the items in the context menu of the variable view.

If there are any items you would like to see added, please comment.

---

Plan 1: `Copy Raw Value`

Convert as follows.

**displayed value**:
```
"line 1`nline 2"
```

**raw value**:
```
line 1
line 2
```

---

Plan 2: `Copy as Binary`, `Copy as Decimal`, `Copy as Hex`

Convert and copy to binary/decimal/hexadecimal respectively.
This may not be used very often, but will be implemented if requested.

**Comment by SAbboushi** [comment]
I need to display value in both decimal and hex very often, usually because AHK displays decimal values and spy tools usually display hwnd in hex

**Reply by zero-plusplus** [reply]
In that usage, it would be nice to have `View as Hex` as shown in the linked image. However, I'm not sure how this is implemented, so I'll implement `Copy as Hex` first.

The `View as Hex` would probably be implemented in a way to show it in a new editor, but it seems a bit over the top for just looking at a single value.

**Reply by SAbboushi** [reply]
If there's a way to display decimal and hex together on hover, that would be my preference, something like:
1,234,567 (0x12 D687)

**Reply by zero-plusplus** [reply]
Unfortunately, there doesn't seem to be an API available to override the hover behavior.

**Reply by zero-plusplus** [reply]
There was no need to override the hover process to solve this problem.
A simple solution would be to add hexadecimal values to the variables.

However, it needs to be optional so as not to confuse beginners in programming.
To do this, I need to come up with an option name that is easy to understand.
Do you have any ideas?

I have the following draft in idea.
```
<launch.json>
...
"numberFormat": "{decimal} (0x{hex})"
...
```

**Reply by SAbboushi** [reply]
1) Does this mean the default will be "numberFormat": "{decimal}"?
2) Will there be a binary option?
3) Any formatting options e.g. commas for decimal, spacing for hex/decimal?

**Reply by zero-plusplus** [reply]
I didn't explain it well enough.

> 1. Does this mean the default will be "numberFormat": "{decimal}"?

Yes, it is. The default is `"{decimal}"`.

> 2. Will there be a binary option?

The four possible values are `binary`, `decimal`, `hex`, and `scientificNotation`.

> 3. Any formatting options e.g. commas for decimal, spacing for hex/decimal?

At the moment, I am thinking that the only above four can be specified.

---

There is one major problem with this, and that is that context menus such as `Copy as Binary` will no longer work.

**Comment by SAbboushi** [comment]
Sometimes I need conversion to/from binary and would be nice to have option to display binary with and without spaces

**Reply by zero-plusplus** [reply]
I'm not familiar with binary, but is a space something like `0 1 0 1`?

**Comment by SAbboushi** [comment]
Yes, but I would prefer spaces on the byte boundaries:
Binary without formatting:                       101001111101011000001000111100110
Binary with formatting (i.e. spaces): 1 0100 1111 1010 1100 0001 0001 1110 0110

**Reply by zero-plusplus** [reply]
Thanks for the explanation.
Does it make any sense that the first 1 in the example with spaces is not 4 digits?

**Comment by SAbboushi** [comment]
To me it does.  I might use leading zeros when aligning numbers on top of each other:
0001 1110 0110
0101 1110 0110

To me, the spaces are a visual aid like commas:
1,234
Commas break numbers into 3 digit segments.  We don't generally write it 001,234

**Reply by zero-plusplus** [reply]
Thank you, I understand.

I'll follow the Windows calculator and implement it as four digits with zeros, like `0001 0101`.

**Reply by zero-plusplus** [reply]
To simplify my work, I decided to rely on Javascript's Number.toString for the conversion process.
Therefore, I will forgo padding the beginning with zeros and adding spaces this time.

**Comment by SAbboushi** [comment]
Windows Calculator seems to ignore leading zeros when formating octal; but includes leading zeros for binary, so I find there is no "standard":
![image](https://user-images.githubusercontent.com/1964198/130324399-18e4e771-493e-4043-aa00-4a827d22cb73.png)

**Comment by zero-plusplus** [comment]
@SAbboushi
I implemented the context menu in `1.10.0`.

See [here](https://github.com/zero-plusplus/vscode-autohotkey-debug/wiki/Data-inspection#context-menu) for details.

**Comment by SAbboushi** [comment]
Great!!
